File: functions/src/shared/growth.ts
// 成長式（企画書 v1.3 §4.3〜§4.5, §5.3）。サーバー側が正。
// tools/growth_ref.py と同一入力で同一出力になることを test/growth_golden.test.ts で保証する。
import { STATS, type GrowthType, type Stat, type Stats, type StatsInt, type TrainingType, loadConfig } from "./config";
import { XorShift128, floorHalfUp } from "./rng";

export interface Individual {
  base: StatsInt;
  talent: StatsInt; // 隠し値。クライアントに返してはならない
  growth: GrowthType; // 隠し値
  personality: number;
  murmurRate: number; // 隠し値
  murmurWindowOffset: number; // 隠し値
}

export function talentFromRaw(rawBase: number, rawTalent: number): number {
  const { constants: C } = loadConfig();
  const v = rawTalent * 0.6 + ((65 - rawBase) / 60) * 10 * 0.4;
  return Math.max(C.talentRange[0], Math.min(C.talentRange[1], floorHalfUp(v)));
}

export function statCap(talent: number): number {
  const { constants: C } = loadConfig();
  return C.statCapBase + talent * C.statCapPerTalent;
}

export function curve(growth: GrowthType, level: number): number {
  const { curves } = loadConfig();
  const v = curves[growth][level - 1];
  if (v === undefined) throw new Error(`curve out of range: ${growth} L${level}`);
  return v;
}

export function rollGrowth(bst0: number, u: number): GrowthType {
  const { constants: C } = loadConfig();
  let weights: number[] | undefined;
  for (const row of C.growthTypeWeights) {
    if (bst0 >= row.minBst0) {
      weights = row.weights;
      break;
    }
  }
  if (!weights) throw new Error("growthTypeWeights has no matching row");
  let cum = 0;
  for (let i = 0; i < C.growthTypeOrder.length; i++) {
    cum += weights[i];
    if (u < cum) return C.growthTypeOrder[i];
  }
  return C.growthTypeOrder[C.growthTypeOrder.length - 1];
}

export function rollIndividual(rng: XorShift128, mentorTalent: StatsInt | null = null, useCapsule = false): Individual {
  const { constants: C, personalities } = loadConfig();
  const base = {} as StatsInt;
  const talent = {} as StatsInt;
  for (const s of STATS) {
    const rb = rng.randInt(C.baseStatRange[0], C.baseStatRange[1]);
    const rt = rng.randInt(C.talentRange[0], C.talentRange[1]);
    base[s] = rb;
    talent[s] = talentFromRaw(rb, rt);
  }
  if (mentorTalent) {
    const rate = useCapsule ? C.mentorInheritRateCapsule : C.mentorInheritRate;
    for (const s of STATS) {
      talent[s] = Math.min(C.talentRange[1], talent[s] + floorHalfUp(mentorTalent[s] * rate));
    }
  }
  const bst0 = STATS.reduce((a, s) => a + base[s], 0);
  const growth = rollGrowth(bst0, rng.nextDouble());
  const personality = rng.randInt(0, personalities.length - 1);
  const murmurRate = rng.randRange(C.murmurRateRange[0], C.murmurRateRange[1]);
  const j = C.murmurWindowJitterLevels;
  const murmurWindowOffset = rng.randInt(-j, j);
  return { base, talent, growth, personality, murmurRate, murmurWindowOffset };
}

/** level → level+1 のレベルアップ後ステータス */
export function levelUp(stats: Stats, talent: StatsInt, growth: GrowthType, level: number, personality: number, rng: XorShift128): Stats {
  const { constants: C, personalities } = loadConfig();
  const mods = personalities[personality].levelGainMod;
  const out = {} as Stats;
  for (const s of STATS) {
    const r = rng.randRange(C.levelGainRandRange[0], C.levelGainRandRange[1]);
    const gain = (C.levelGainBase + C.levelGainPerTalent * talent[s]) * curve(growth, level) * mods[s] * r;
    out[s] = Math.min(statCap(talent[s]), stats[s] + gain);
  }
  return out;
}

export function trainingMod(personality: number, type: TrainingType): number {
  const { constants: C, personalities } = loadConfig();
  const p = personalities[personality];
  if (p.likes === type) return 1.0 + C.personalityTrainingBonus;
  if (p.dislikes === type) return 1.0 - C.personalityTrainingBonus;
  return 1.0;
}

export function train(
  stats: Stats,
  talent: StatsInt,
  personality: number,
  type: TrainingType,
  fatigue: number,
  rng: XorShift128,
): { stats: Stats; fatigue: number } {
  const { constants: C } = loadConfig();
  const t = C.trainings[type];
  const main: Stat = t.main;
  const mainGain = (rng.randRange(C.trainingGainRange[0], C.trainingGainRange[1]) + C.trainingGainPerTalent * talent[main]) * trainingMod(personality, type);
  const out = { ...stats };
  out[main] = Math.min(statCap(talent[main]), stats[main] + mainGain);
  if (t.sub) {
    out[t.sub] = Math.min(statCap(talent[t.sub]), stats[t.sub] + mainGain * C.trainingSubRatio);
  }
  return { stats: out, fatigue: fatigue + t.fatigue };
}

export function expToNext(level: number): number {
  const { constants: C } = loadConfig();
  return C.expBase + C.expPerLevel * level;
}

export function applyExp(level: number, exp: number, gained: number): { level: number; exp: number; levelUps: number } {
  const { constants: C } = loadConfig();
  const cap = C.levelCap;
  if (level >= cap) return { level: cap, exp: 0, levelUps: 0 };
  let levelUps = 0;
  exp += gained;
  while (level < cap && exp >= expToNext(level)) {
    exp -= expToNext(level);
    level += 1;
    levelUps += 1;
  }
  if (level >= cap) exp = 0;
  return { level, exp, levelUps };
}

File: app/lib/domain/growth.dart
// 成長式（企画書 v1.3 §4.3〜§4.5, §5.3）。
// サーバー側（functions/src/shared/growth.ts）が正。クライアントは表示用の予測にのみ使う。
// tools/growth_ref.py と同一入力で同一出力になることを test/growth_golden_test.dart で保証する。
// 純 Dart（Flutter 非依存）。
import 'rng.dart';

const List<String> kStats = ['hp', 'atk', 'def', 'spa', 'spd', 'luk'];

/// shared-config の constants.json / personalities.json / growth_curves.json を束ねたもの。
/// アプリ層が assets から読んで組み立てる。
class GameConfig {
  GameConfig({
    required this.constants,
    required this.personalities,
    required this.curves,
  });

  final Map<String, dynamic> constants;
  final List<dynamic> personalities;
  final Map<String, dynamic> curves;

  int get levelCap => constants['levelCap'] as int;
  List<String> get growthTypeOrder => (constants['growthTypeOrder'] as List).cast<String>();

  num _n(String key) => constants[key] as num;
  List<num> _range(String key) => (constants[key] as List).cast<num>();
}

class Individual {
  Individual({
    required this.base,
    required this.talent,
    required this.growth,
    required this.personality,
    required this.murmurRate,
    required this.murmurWindowOffset,
  });

  final Map<String, int> base;

  /// 隠し値。本番クライアントはこれを受け取らない（表示用シミュレーションのみ）。
  final Map<String, int> talent;
  final String growth;
  final int personality;
  final double murmurRate;
  final int murmurWindowOffset;
}

int talentFromRaw(GameConfig cfg, int rawBase, int rawTalent) {
  final v = rawTalent * 0.6 + (65 - rawBase) / 60 * 10 * 0.4;
  final r = cfg._range('talentRange');
  return floorHalfUp(v).clamp(r[0].toInt(), r[1].toInt());
}

int statCap(GameConfig cfg, int talent) =>
    cfg._n('statCapBase').toInt() + talent * cfg._n('statCapPerTalent').toInt();

double curve(GameConfig cfg, String growth, int level) {
  final table = (cfg.curves[growth] as List).cast<num>();
  return table[level - 1].toDouble();
}

String rollGrowth(GameConfig cfg, int bst0, double u) {
  List<num>? weights;
  for (final row in cfg.constants['growthTypeWeights'] as List) {
    if (bst0 >= (row['minBst0'] as num)) {
      weights = (row['weights'] as List).cast<num>();
      break;
    }
  }
  if (weights == null) throw StateError('growthTypeWeights has no matching row');
  final order = cfg.growthTypeOrder;
  var cum = 0.0;
  for (var i = 0; i < order.length; i++) {
    cum += weights[i];
    if (u < cum) return order[i];
  }
  return order.last;
}

Individual rollIndividual(GameConfig cfg, XorShift128 rng,
    {Map<String, int>? mentorTalent, bool useCapsule = false}) {
  final baseR = cfg._range('baseStatRange');
  final talR = cfg._range('talentRange');
  final base = <String, int>{};
  final talent = <String, int>{};
  for (final s in kStats) {
    final rb = rng.randInt(baseR[0].toInt(), baseR[1].toInt());
    final rt = rng.randInt(talR[0].toInt(), talR[1].toInt());
    base[s] = rb;
    talent[s] = talentFromRaw(cfg, rb, rt);
  }
  if (mentorTalent != null) {
    final rate = (useCapsule ? cfg._n('mentorInheritRateCapsule') : cfg._n('mentorInheritRate')).toDouble();
    for (final s in kStats) {
      final t = talent[s]! + floorHalfUp(mentorTalent[s]! * rate);
      talent[s] = t < talR[1].toInt() ? t : talR[1].toInt();
    }
  }
  final bst0 = kStats.fold<int>(0, (a, s) => a + base[s]!);
  final growth = rollGrowth(cfg, bst0, rng.nextDouble());
  final personality = rng.randInt(0, cfg.personalities.length - 1);
  final mr = cfg._range('murmurRateRange');
  final murmurRate = rng.randRange(mr[0].toDouble(), mr[1].toDouble());
  final j = cfg._n('murmurWindowJitterLevels').toInt();
  final murmurWindowOffset = rng.randInt(-j, j);
  return Individual(
    base: base,
    talent: talent,
    growth: growth,
    personality: personality,
    murmurRate: murmurRate,
    murmurWindowOffset: murmurWindowOffset,
  );
}

/// level → level+1 のレベルアップ後ステータス。
Map<String, double> levelUp(GameConfig cfg, Map<String, double> stats, Map<String, int> talent,
    String growth, int level, int personality, XorShift128 rng) {
  final mods = (cfg.personalities[personality]['levelGainMod'] as Map).cast<String, num>();
  final rr = cfg._range('levelGainRandRange');
  final gainBase = cfg._n('levelGainBase').toDouble();
  final gainPerTalent = cfg._n('levelGainPerTalent').toDouble();
  final out = <String, double>{};
  for (final s in kStats) {
    final r = rng.randRange(rr[0].toDouble(), rr[1].toDouble());
    final gain = (gainBase + gainPerTalent * talent[s]!) * curve(cfg, growth, level) * mods[s]!.toDouble() * r;
    final cap = statCap(cfg, talent[s]!).toDouble();
    final v = stats[s]! + gain;
    out[s] = v < cap ? v : cap;
  }
  return out;
}

double trainingMod(GameConfig cfg, int personality, String type) {
  final p = cfg.personalities[personality] as Map;
  final b = cfg._n('personalityTrainingBonus').toDouble();
  if (p['likes'] == type) return 1.0 + b;
  if (p['dislikes'] == type) return 1.0 - b;
  return 1.0;
}

class TrainResult {
  TrainResult(this.stats, this.fatigue);
  final Map<String, double> stats;
  final int fatigue;
}

TrainResult train(GameConfig cfg, Map<String, double> stats, Map<String, int> talent, int personality,
    String type, int fatigue, XorShift128 rng) {
  final t = (cfg.constants['trainings'] as Map)[type] as Map;
  final main = t['main'] as String;
  final sub = t['sub'] as String?;
  final gr = cfg._range('trainingGainRange');
  final mainGain = (rng.randRange(gr[0].toDouble(), gr[1].toDouble()) +
          cfg._n('trainingGainPerTalent').toDouble() * talent[main]!) *
      trainingMod(cfg, personality, type);
  final out = Map<String, double>.from(stats);
  final capMain = statCap(cfg, talent[main]!).toDouble();
  final vm = stats[main]! + mainGain;
  out[main] = vm < capMain ? vm : capMain;
  if (sub != null) {
    final capSub = statCap(cfg, talent[sub]!).toDouble();
    final vs = stats[sub]! + mainGain * cfg._n('trainingSubRatio').toDouble();
    out[sub] = vs < capSub ? vs : capSub;
  }
  return TrainResult(out, fatigue + (t['fatigue'] as num).toInt());
}

int expToNext(GameConfig cfg, int level) =>
    cfg._n('expBase').toInt() + cfg._n('expPerLevel').toInt() * level;

class ExpResult {
  ExpResult(this.level, this.exp, this.levelUps);
  final int level;
  final int exp;
  final int levelUps;
}

ExpResult applyExp(GameConfig cfg, int level, int exp, int gained) {
  final cap = cfg.levelCap;
  if (level >= cap) return ExpResult(cap, 0, 0);
  var levelUps = 0;
  exp += gained;
  while (level < cap && exp >= expToNext(cfg, level)) {
    exp -= expToNext(cfg, level);
    level += 1;
    levelUps += 1;
  }
  if (level >= cap) exp = 0;
  return ExpResult(level, exp, levelUps);
}
